core/src/main/scala/org/apache/spark/util/JsonProtocol.scala

```diff
@@ -663,66 +854,69 @@ private[spark] object JsonProtocol {
+    val stageInfo = stageInfoFromJson(json.get("Stage Info"))
+    val properties = propertiesFromJson(json.get("Properties"))
     SparkListenerStageSubmitted(stageInfo, properties)
   }
 
-  def stageCompletedFromJson(json: JValue): SparkListenerStageCompleted = {
-    val stageInfo = stageInfoFromJson(json \ "Stage Info")
+  def stageCompletedFromJson(json: JsonNode): SparkListenerStageCompleted = {
+    val stageInfo = stageInfoFromJson(json.get("Stage Info"))
     SparkListenerStageCompleted(stageInfo)
   }
 
-  def taskStartFromJson(json: JValue): SparkListenerTaskStart = {
-    val stageId = (json \ "Stage ID").extract[Int]
+  def taskStartFromJson(json: JsonNode): SparkListenerTaskStart = {
+    val stageId = json.get("Stage ID").intValue
     val stageAttemptId =
-      jsonOption(json \ "Stage Attempt ID").map(_.extract[Int]).getOrElse(0)
-    val taskInfo = taskInfoFromJson(json \ "Task Info")
+      jsonOption(json.get("Stage Attempt ID")).map(_.intValue).getOrElse(0)
+    val taskInfo = taskInfoFromJson(json.get("Task Info"))
     SparkListenerTaskStart(stageId, stageAttemptId, taskInfo)
   }
 
-  def taskGettingResultFromJson(json: JValue): SparkListenerTaskGettingResult = {
-    val taskInfo = taskInfoFromJson(json \ "Task Info")
+  def taskGettingResultFromJson(json: JsonNode): SparkListenerTaskGettingResult = {
+    val taskInfo = taskInfoFromJson(json.get("Task Info"))
     SparkListenerTaskGettingResult(taskInfo)
   }
 
   /** Extract the executor metrics from JSON. */
-  def executorMetricsFromJson(json: JValue): ExecutorMetrics = {
+  def executorMetricsFromJson(maybeJson: JsonNode): ExecutorMetrics = {
+    // Executor metrics might be absent in JSON from very old Spark versions.
+    // In this case we return zero values for each metric.
     val metrics =
       ExecutorMetricType.metricToOffset.map { case (metric, _) =>
-        metric -> jsonOption(json \ metric).map(_.extract[Long]).getOrElse(0L)
+        val metricValueJson = jsonOption(maybeJson).flatMap(json => jsonOption(json.get(metric)))
+        metric -> metricValueJson.map(_.longValue).getOrElse(0L)
       }
     new ExecutorMetrics(metrics.toMap)
   }
 
-  def taskEndFromJson(json: JValue): SparkListenerTaskEnd = {
-    val stageId = (json \ "Stage ID").extract[Int]
+  def taskEndFromJson(json: JsonNode): SparkListenerTaskEnd = {
+    val stageId = json.get("Stage ID").intValue
     val stageAttemptId =
-      jsonOption(json \ "Stage Attempt ID").map(_.extract[Int]).getOrElse(0)
-    val taskType = (json \ "Task Type").extract[String]
-    val taskEndReason = taskEndReasonFromJson(json \ "Task End Reason")
-    val taskInfo = taskInfoFromJson(json \ "Task Info")
-    val executorMetrics = executorMetricsFromJson(json \ "Task Executor Metrics")
-    val taskMetrics = taskMetricsFromJson(json \ "Task Metrics")
+      jsonOption(json.get("Stage Attempt ID")).map(_.intValue).getOrElse(0)
```

Comment:
At the start of this work, I did some toy experiments to see if there were potential behavior differences between Jackson and Json4s when it came to handilng unexpected types (e.g. a field which you expect to contain an int but which contains a string that is an int value, or vice-versa). I think that all of the cases were handled in the same way, but I'd like to reconfirm this.

I'll repeat those experiments and will share the results here.

Reply:
I did some experiments in a notebook with code similar to this:

```scala
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.json4s.JsonDSL._
import org.json4s.DefaultFormats

import com.fasterxml.jackson.core.{JsonEncoding, JsonGenerator}
import com.fasterxml.jackson.databind.{DeserializationFeature, JsonNode, ObjectMapper}
import com.fasterxml.jackson.module.scala.DefaultScalaModule

implicit val format = DefaultFormats

private val mapper = new ObjectMapper().registerModule(DefaultScalaModule)
  .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
```

and test cases like

```scala
import scala.util.Try
val json = """{"value": 21474836470}"""
val json4sResult = Try((parse(json) \ "value").extract[Int])
val jacksonResult = Try(mapper.readTree(json).get("value").intValue)
assert(json4sResult == jacksonResult, s"json4s: ${json4sResult}, jackson: ${jacksonResult}")
```

This uncovered one subtle corner-case: if you have a string field where a number is expected then my new code will silently return the zero value for the numeric type, whereas the old code would have failed with an exception:

```scala
import scala.util.Try
val json = """{"value": "1"}"""
val json4sResult = Try((parse(json) \ "value").extract[Long])
val jacksonResult = Try(mapper.readTree(json).get("value").longValue)
assert(json4sResult == jacksonResult, s"json4s: ${json4sResult}, jackson: ${jacksonResult}")
// outputs: AssertionError: assertion failed: json4s: Failure(org.json4s.MappingException: Do not know how to convert JString(1) into long), jackson: Success(0)
```

Similarly, `.textValue` will return `null` for non-Strings:

```scala
import scala.util.Try
val json = """{"value":1}"""
val json4sResult = Try((parse(json) \ "value").extract[String])
val jacksonResult = Try(mapper.readTree(json).get("value").textValue)
assert(json4sResult == jacksonResult, s"json4s: ${json4sResult}, jackson: ${jacksonResult}")
// outputs: AssertionError: assertion failed: json4s: Success(1), jackson: Success(null)
```

These behavior differences aren't an issue if we assume that Spark is always parsing well-formed JSON generated by itself,  but it is a potential risk of silent and confusing behaviors in case Spark reads malformed JSON. Given this, I'm going to explore whether I can write helper functions to add type checking so as to match the old behavior.

Reply:
Pushed 03555c6f6074a87c64c059eeddf665b9ffc69c9a to add implicit `extract*` methods that perform typechecking. This has maybe a _slight_ performance impact but the new parsing code is still ~2x faster than the old Json4s code in my local tests.